GVFS/GVFS.Common/Http/HttpRequestor.cs

```diff
@@ -249,10 +316,69 @@ private static string GetSingleHeaderOrEmpty(HttpHeaders headers, string headerN
             IEnumerable<string> values;
             if (headers.TryGetValues(headerName, out values))
             {
-                return values.First();                
+                return values.First();
             }
 
             return string.Empty;
         }
+
+        private string LoadCertificatePassword(string certId, GitProcess git)
+        {
+            if (git.TryGetCertificatePassword(this.Tracer, certId, out var password, out var error))
+            {
+                return password;
+            }
+
+            return null;
+        }
+
+        private X509Certificate2 LoadCertificate(string certId, string certificatePassword, bool onlyLoadValidCertificateFromStore)
+        {
+            EventMetadata metadata = new EventMetadata
+            {
+                { "certId", certId },
+                { "isPasswordSpecified", string.IsNullOrEmpty(certificatePassword) },
+                { "shouldVerify", onlyLoadValidCertificateFromStore }
+            };
+
+            if (File.Exists(certId))
+            {
+                try
+                {
+                    var cert = new X509Certificate2(certId, certificatePassword);
+                    if (onlyLoadValidCertificateFromStore && cert != null && !cert.Verify())
+                    {
+                        this.Tracer.RelatedWarning(metadata, "Certficate was found, but is invalid.");
+                        return null;
+                    }
+
+                    return cert;
+                }
+                catch (CryptographicException cryptEx)
+                {
+                    metadata.Add("Exception", cryptEx);
+                    this.Tracer.RelatedError(metadata, "Error, while loading certificate from disk");
+                    return null;
+                }
+            }
+
+            try
+            {
+                var findResults = this.store.Value.Certificates.Find(X509FindType.FindBySubjectName, certId, onlyLoadValidCertificateFromStore);
+                if (findResults?.Count > 0)

```

Comment:
If there is more than 1, do we care?  How do we know we are picking the right one?

Reply:
There is another option of:
3. Try each one until one works or fail if none work, so add a foreach loop.

Reading up on this and don't we want to use [`FindBySubjectDistinguishedName`](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509findtype?view=netframework-4.7.2)?  The documentation says that `FindBySubjectName`
>will find all certificates with the subject name containing that string, regardless of other subject values. Searching by distinguished name is a more precise search.

To me that mean that if I pass "test", I will get the certificates for "mytest", "test", "testing", etc.  

So option 3 would be the only one to work in that case.  Do we have the full name of the cert so we can use `FindBySubjectDistinguishedName`?

Reply:
I like your plan of logging a warning and narrowing down the certs.  Like you said hopefully there will be enough information in the logs to find out what is going on.  Thanks

Reply:
Pushed changes, implementing the above.